gateway/mw_granular_access.go

```diff
@@ -34,20 +33,30 @@ func (m *GranularAccessMiddleware) ProcessRequest(w http.ResponseWriter, r *http
 		return nil, http.StatusOK
 	}
 
-	urlPath := m.Spec.StripListenPath(r.URL.Path)
+	urlPaths := []string{
+		m.Spec.StripListenPath(r.URL.Path),
+		r.URL.Path,
+	}
+
+	logger := m.Logger().WithField("path", r.URL.Path)
 
 	for _, accessSpec := range sessionVersionData.AllowedURLs {
 		url := accessSpec.URL
-		match, err := httputil.MatchEndpoint(url, urlPath)
-		if err != nil {
-			logger.WithError(err).Errorf("error matching path regex: %q, skipping", url)
-			continue
+
+		var match bool
+		var err error
+		for _, urlPath := range urlPaths {
+			match, err = httputil.MatchEndpoint(url, urlPath)
+			if err != nil || !match {
+				logger.WithError(err).Errorf("error matching path regex: %q, skipping", url)
+				continue
+			}
+			break
```

Comment:
@titpetric does PR has a full changelist? I feel this break is not described there. What it does?

Reply:
@buger this break is introduced to do multiple matches and reuse match/err;

- L36 sets input url paths,
- L48 loops them, match/err values are reused,
- if a match is not found or an error occured it will use the next value (log and continue)
- if a match is found it will break and continue to L57 where the final error checks happen (no urlPaths matched) and goes to the next allowed url in the list;

Only the url stripped of the listenpath would be considered since PR #6437 (parent ticket); before that, only r.URL.Path would be considered, ticket and PR description explain.

Reply:
Moved the logic into httputil.MatchEndpoints(string, []string), so it's not adding to the cognitive complexity. It should be clear, covered with godoc.